akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceWithContextSpec.scala

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2014-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.scaladsl
+
+import akka.stream.testkit.scaladsl.TestSink
+import akka.stream.{ ActorMaterializer, ActorMaterializerSettings }
+import akka.stream.testkit.StreamSpec
+
+case class Message(data: String, offset: Long)
+
+class SourceWithContextSpec extends StreamSpec {
+
+  val settings = ActorMaterializerSettings(system)
+  implicit val materializer = ActorMaterializer(settings)
+
+  "A SourceWithContext" must {
+
+    "get created from Source.withContext" in {
+      val msg = Message("a", 1L)
+      Source(Vector(msg))
+        .withContext
+        .provideContext
+        .runWith(TestSink.probe[(Message, Op[Message])])
+        .request(1)
+        .expectNext((msg, OneToOne(msg)))
+        .expectComplete()
+    }
+
+    "turn back into a Source using withoutContext" in {
+      val msg = Message("a", 1L)
+      Source(Vector(msg))
+        .withContext
+        .map(_.data)
+        .withoutContext
+        .runWith(TestSink.probe[String])
+        .request(1)
+        .expectNext("a")
+        .expectComplete()
+    }
+
+    "be able to provideNoOp" in {
+      val msg = Message("a", 1L)
+      val srcCtx = Source(Vector(msg)).withContext
+
+      val normalSrc = srcCtx.withoutContext.map(_.data)
+
+      normalSrc.withContext.provideNoOp
+        .runWith(TestSink.probe[(String, Op[Nothing])])
+        .request(1)
+        .expectNext(("a", NoOp))
+        .expectComplete()
+    }
+
+    "pass through a one-to-one contextual operation" in {
+      Source(
+        Vector(Message("A", 1L), Message("B", 2L), Message("D", 3L), Message("C", 4L))
+      )
+        .withContext
+        .mapContext(_.offset)
+        .map(_.data.toLowerCase)
+        .filter(_ != "b")
+        .filterNot(_ == "d")
+        .provideContext
+        .runWith(TestSink.probe[(String, Op[Long])])
+        .request(2)
+        .expectNext(("a", OneToOne(1L)))
+        .expectNext(("c", OneToOne(4L)))
+        .expectComplete()
+    }
+
+    "provide a one-to-one contextual operation via a FlowWithContext" in {
+
+      def flowWithContext[T] = FlowWithContext[Long, T]
+
+      Source(Vector(Message("a", 1L)))
+        .withContext
+        .mapContext(_.offset)
+        .map(_.data)
+        .via(flowWithContext.map(s ⇒ s + "b"))
+        .provideContext
```

Comment:
When just looking at the API I find the name `provideContext` misleading. I though `provide` was that the user should "provide" what the context was, but it's about converting the context aware stream back to a normal stream?

Reply:
Would be nice with some naming symmetry between `withContext` and `provideContext`.

```
.startContextPropagation
.mapContext
.endContextPropagation
```

Reply:
done.